modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,234 @@
+        :return: The mode of the pin.
+        """
+        self._check_if_disabled()
+        return self._mode
+
+    @mode.setter
+    def mode(self, value: GPIOPinMode) -> None:
+        """
+        Set the mode of the pin.
+
+        To do analogue or digital reads set the mode to INPUT or INPUT_PULLUP.
+        To do digital writes set the mode to OUTPUT.
+
+        :param value: The mode to set the pin to.
+        :raises IOError: If the pin mode is not a GPIOPinMode.
+        :raises IOError: If this pin cannot be controlled.
+        """
+        self._check_if_disabled()
+        if not isinstance(value, GPIOPinMode):
+            raise IOError('Pin mode only supports being set to a GPIOPinMode')
+
+        self._mode = value
+
+    def _digital_read(self) -> bool:
+        return self._analogue_read() > 1
+
+    def digital_read(self) -> bool:
+        """
+        Perform a digital read on the pin.
+
+        :raises IOError: If the pin's current mode does not support digital read
+        :raises IOError: If this pin cannot be controlled.
+        :return: The digital value of the pin.
+        """
+        self._check_if_disabled()
+        if self.mode not in DIGITAL_READ_MODES:
+            raise IOError(f'Digital read is not supported in {self.mode}')
+        return self._digital_read()
+
+    def _digital_write(self, value: bool) -> None:
+        return
+
+    def digital_write(self, value: bool) -> None:
+        """
+        Write a digital value to the pin.
+
+        :param value: The value to write to the pin.
+        :raises IOError: If the pin's current mode does not support digital write.
+        :raises IOError: If this pin cannot be controlled.
+        """
+        self._check_if_disabled()
+        if self.mode not in DIGITAL_WRITE_MODES:
+            raise IOError(f'Digital write is not supported in {self.mode}')
+
+        self._digital_write(value)
+
+    def _analogue_read(self) -> float:
+        return map_to_range(0, 1, *self._ANALOGUE_RANGE, random.random())
+
+    def analog_read(self) -> float:
+        """
+        Get the analogue voltage on the pin.
+
+        This is returned in volts. Only pins A0-A7 support analogue reads.
+
+        :raises IOError: If the pin or its current mode does not support analogue read.
+        :raises IOError: If this pin cannot be controlled.
+        :return: The analogue voltage on the pin, ranges from 0 to 5.
+        """
+        self._check_if_disabled()
+        if self.mode not in ANALOG_READ_MODES:
+            raise IOError(f'Analogue read is not supported in {self.mode}')
+        if not self._supports_analogue:
+            raise IOError('Pin does not support analogue read')
+
+        return add_jitter(self._analogue_read(), *self._ANALOGUE_RANGE)
+
+
+class DisabledPin(PinDevice):
+    def __init__(self) -> None:
+        super().__init__(supports_analogue=False, disabled=True)
+
+
+class EmptyPin(PinDevice):
+    pass
```

Comment:
These subclasses seem unnecessary, just use the base class.

Reply:
These classes allow us to more clearly express intent of combinations of features.

Reply:
0d4f0fb930bae52e020ab2feed3be1d206479229